core/services/workflows/engine.go

```diff
@@ -19,100 +22,143 @@ const (
 	mockedTriggerID   = "cccccccccc0000000000000000000000"
 )
 
+// Engine handles the lifecycle of a single workflow and its executions.
 type Engine struct {
 	services.StateMachine
-	logger     logger.Logger
-	registry   types.CapabilitiesRegistry
-	trigger    capabilities.TriggerCapability
-	consensus  capabilities.ConsensusCapability
-	targets    []capabilities.TargetCapability
-	workflow   *Workflow
-	callbackCh chan capabilities.CapabilityResponse
-	cancel     func()
+	logger              logger.Logger
+	registry            types.CapabilitiesRegistry
+	workflow            *workflow
+	executionStates     *inMemoryStore
+	pendingStepRequests chan stepRequest
+	triggerEvents       chan capabilities.CapabilityResponse
+	newWorkerCh         chan struct{}
+	stepUpdateCh        chan stepState
+	wg                  sync.WaitGroup
+	stopCh              services.StopChan
+	newWorkerTimeout    time.Duration
+
+	// Used for testing to wait for an execution to complete
+	xxxExecutionFinished chan string
```

Comment:
Not a huge fan of these kind of hooks. Can we maybe use a mock target in tests and wait for it to be called?

Reply:
Same... I initially used a mock target but that leads to a race since the engine will do some processing (like updating internal state) after the target has been called. It also won't notify you if the workflow has errored for some reason.